ls1intum/Artemis
# `Quiz exercises`: Add functionality to import existing quiz

### Checklist
#### General


#### Server
#### Client

Closes #4999 

### Motivation and Context



Adds the possibility of importing an already existing quiz, therefore allowing instructors to already use the quiz/questions from a previous course without having to start from scratch.

### Description


In the client, added a new "Import quiz" button that opens an import modal showing all existing quizzes from other courses/exams that the user has access to. The user can search for quizzes by their title or course name. The user can then click on the "Import" button and the "edit quiz" page will open pre-filled with the values/questions of the selected quiz. The user can change any field/question as desired. When the "Save" button is clicked, a new quiz will be imported/created.

In the server, added a new REST endpoint that works similar to the createQuiz endpoint, except it uses an already existing quiz as a template.

Used the similar "Import Text Exercise" and "Import Modeling Exercise" functionality as template and emulated the behavior in the Client and server.

### Steps for Testing

Prerequisites:
- 1 Instructor
- (at least) 1 Course with (at least) 1 quiz exercise

1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to the exercises for a course
4. Click the "Import quiz" button
5. Search for the quiz you want to import and click the "Import" button
6. Verify that an "edit" page for a quiz is opened, pre-filled with the values of the existing quiz
7. Verify that all date related values have been reset
8. Change any values/questions as desired
9. Click the "Save" button
10. Verify that you were taken back to the quiz exercises page and that a new quiz exercise has been created
11. Click the "Edit" button for the created quiz and verify that all values/questions were set/created correctly
12. You can follow steps 4-11 but for an exam exercise group instead of a course

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| QuizExerciseResource.java | 80% | 93% |
| QuizExerciseImportService.java | 82% | 97% |
| QuizExerciseRepository.java | na | 90% |
| quiz-exercise-detail.component.ts | 82% | 96% |
| quiz-exercise-import.component.ts | 76% | 95% |
| quiz-exercise-paging.service.ts | 83% | 100% |

### Screenshots



Course management exercises page with new "Import quiz" button

![Import quiz for course exercise](https://user-images.githubusercontent.com/11006274/169717773-3d78ae05-2a46-4940-b7ff-88d726878668.png)

Exam exercise group exercises page with new "Import quiz" button

![Import quiz for exam](https://user-images.githubusercontent.com/11006274/169717777-8b41fe61-15f9-4198-8c37-81908347e035.png)

Open import modal showing all existing quizzes that can be imported

![New Quiz import modal](https://user-images.githubusercontent.com/11006274/169717780-bf4bb1f0-fef9-4542-b76a-7165b2d799b8.png)
